fetchenvoy/main.go

```diff
@@ -0,0 +1,226 @@
+package main
+
+import (
+	"archive/tar"
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"os/exec"
+	"time"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/client"
+	specs "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/spf13/cobra"
+	"go.uber.org/zap"
+)
+
+const envoyRepo = "envoyproxy/envoy"
+const envoyPath = "/usr/local/bin/envoy"
+
+var envoyArchs = []string{"arm64", "amd64"}
+
+var log *zap.SugaredLogger
+
+func init() {
+	l, err := zap.NewDevelopment()
+	log = l.Sugar()
+	if err != nil {
+		panic(err)
+	}
+}
+
+var rootCmd = &cobra.Command{
+	Use:   "fetchenvoy [image tag]",
+	Short: "fetch envoy binaries from upstream containers",
+	Args:  cobra.ExactArgs(1),
+	RunE: func(cmd *cobra.Command, args []string) error {
+		err := run(cmd.Context(), args[0])
+		if err != nil {
+			log.Errorw("failed", "error", err)
+		}
+		return err
+	},
+}
+
+func main() {
+	defer log.Sync() //nolint: errcheck
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute*5)
+	defer cancel()
+
+	err := rootCmd.ExecuteContext(ctx)
+	if err != nil {
+		os.Exit(1)
+	}
+}
+
+func run(ctx context.Context, tag string) error {
+	log.Infow("fetching envoy binaries from docker images", "version", tag, "repo", envoyRepo)
+	client, err := client.NewClientWithOpts(client.WithAPIVersionNegotiation())
+	if err != nil {
+		return fmt.Errorf("client error: %w", err)
+	}
+
+	var files []string
+	for _, arch := range envoyArchs {
+		// start a dummy container to grab binary from
+		log.Infow("processing image", "arch", arch)
+		id, err := doContainer(ctx, client, arch, tag)
+		if err != nil {
+			return err
+		}
+
+		// copy the binary out
+		outFile := fmt.Sprintf("envoy-linux-%s", arch)
+		err = doGetEnvoy(ctx, client, id, outFile)
+		if err != nil {
+			return err
+		}
+
+		// clean up
+		err = client.ContainerRemove(ctx, id, types.ContainerRemoveOptions{})
```

Comment:
maybe should also call this right before line 79?

Reply:
Good call.  Added an attempt at cleanup if we fail earlier.